### PR TITLE
Add GCP PubSub Sink

### PR DESCRIPTION
**Pull request by k-gupta**
Add Google Cloud Platform's Pub Sub Sink

### CODE REVIEW

File: packages/gcp/src/config.ts [hunk 1]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+        return config.noop();
+    }
+    @config.field(config.converters.number)
+    public set maximumBatchSize(_: number) {
+        config.noop();
+    }
+    public get maximumBatchSize(): number {
+        return config.noop();
+    }
+    @config.field(config.converters.number)
```
Comment (sklose) [inline packages/gcp/src/config.ts:137]: might wanna use timespan converter here ... also maybe rename it, `maxMilliseconds` isn't quite self explanatory as a top-level config field by itself.

File: packages/gcp/src/config.ts [hunk 2]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+    public set maxMilliseconds(_: number) {
+        config.noop();
+    }
+    public get maxMilliseconds(): number {
+        return config.noop();
+    }
+    @config.field(config.converters.number)
+    public set maxBytes(_: number) {
```
Comment (sklose) [inline packages/gcp/src/config.ts:145]: there is a converter for bytes as well

File: packages/gcp/src/__test__/PubSubSink.test.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        expect(mockPublishFn).toBeCalledTimes(messages.length);
+        messages.forEach((message, idx) => {
+            expect(mockPublishFn.mock.calls[idx][1][AttributeNames.eventType]).toBe(message.type);
+            expect(mockPublishFn.mock.calls[idx][1][AttributeNames.contentType]).toBe(
+                pubSubPublisherConfigurationWithDefaultTopic.encoder.mimeType
+            );
+        });
+    });
+
+    it("Reject on error from PubSub topic", async () => {
```
Comment (sklose) [inline packages/gcp/src/__test__/PubSubSink.test.ts:132]: nit: can you rename the test cases so they read like a sentence starting with `it`? e.g.

```
test("PubSub Sink") {
  it("writes messages to topics") {
      ...
  }
}
```

File: packages/gcp/package.json [hunk 4]
```diff
@@ -13,8 +13,9 @@
         "url": "https://github.com/walmartlabs/cookie-cutter/issues"
     },
     "dependencies": {
-        "@google-cloud/storage": "4.4.0",
-        "@google-cloud/bigquery": "4.7.0"
+        "@google-cloud/bigquery": "4.7.0",
+        "@google-cloud/pubsub": "2.0.0",
```
Comment (sklose) [inline packages/gcp/package.json:17]: i think you are missing the yarnlock file as part of this PR ... i'd expect the pubsub lib to have been added there

Comment (k-gupta) [reply packages/gcp/package.json:17]: Added, I was missing the step to run the yarn build

File: packages/gcp/src/PubSubSink.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        span.log({ topic });
+        span.setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_RPC_CLIENT);
+        span.setTag(Tags.COMPONENT, "cookie-cutter-pubSub");
+        span.setTag(OpenTracingTagKeys.FunctionName, funcName);
+        span.setTag(PubSubOpenTracingTagKeys.TopicName, topic);
+    }
+
+    private emitMetrics(topic: string, eventType: string, result: PubSubMetricResults) {
+        this.metrics.increment(PubSubMetrics.MsgPublished, {
+            topic: topic,
```
Comment (plameniv) [inline packages/gcp/src/PubSubSink.ts:94]: replace `topic: topic` with just `topic`

File: packages/gcp/src/config.ts (hunk 1 above)
Comment (k-gupta) [reply packages/gcp/src/config.ts:137]: Yup can do, I copied it from the GCP library itself.

Comment (k-gupta) [reply packages/gcp/src/config.ts]: Donee

File: packages/gcp/src/config.ts (hunk 2 above)
Comment (k-gupta) [reply packages/gcp/src/config.ts]: Done

File: packages/gcp/src/PubSubSink.ts (hunk 5 above)
Comment (k-gupta) [reply packages/gcp/src/PubSubSink.ts]: Done

File: packages/gcp/src/__test__/PubSubSink.test.ts [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        mockTopic.mockImplementation(() => mockFns);
+        mockPubSub.mockImplementation(() => mockFns);
+    });
+
+    it("Sink writes to default topic in pubsub", async () => {
+        const sink = new PubSubSink({
+            ...gcsAuthConfig,
+            ...pubSubPublisherConfigurationWithDefaultTopic,
+        });
+        await sink.initialize(DefaultComponentContext);
```
Comment (sklose) [inline packages/gcp/src/__test__/PubSubSink.test.ts:91]: mind writing these tests with a full CC Application and a StaticInputSource? that tends to be less brittle than testing a sink directly

File: packages/gcp/src/config.ts [hunk 7]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+        return config.noop();
+    }
+    @config.field(config.converters.timespan)
+    public set maximumBatchWaitTime(_: number) {
+        config.noop();
+    }
+    public get maximumBatchWaitTime(): number {
+        return config.noop();
+    }
+    @config.field(config.converters.number)
```
Comment (sklose) [inline packages/gcp/src/config.ts:144]: there is a converter for bytes as well

Comment (k-gupta) [reply packages/gcp/src/config.ts:144]: Oh missed this one

File: packages/gcp/src/__test__/PubSubSink.test.ts (hunk 6 above)
Comment (k-gupta) [reply packages/gcp/src/__test__/PubSubSink.test.ts:91]: Yup can do

File: packages/gcp/src/config.ts [hunk 8]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+        return config.noop();
+    }
+    @config.field(config.converters.timespan)
+    public set maximumBatchWaitTime(_: number) {
+        config.noop();
+    }
+    public get maximumBatchWaitTime(): number {
+        return config.noop();
+    }
+    @config.field(config.converters.bytes)
```
Comment (chandanRIT) [inline packages/gcp/src/config.ts:144]: shouldnt this be number?

File: packages/gcp/src/config.ts [hunk 9]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+@config.section
+export class PubSubPublisherConfiguration extends GcpAuthConfiguration
+    implements IPubSubPublisherConfiguration {
+    @config.field(config.converters.none)
+    public set encoder(_: IMessageEncoder) {
+        config.noop();
+    }
+    public get encoder(): IMessageEncoder {
+        return config.noop();
+    }
```
Comment (chandanRIT) [inline packages/gcp/src/config.ts:122]: newline between methods

File: packages/gcp/src/config.ts [hunk 10]
```diff
@@ -77,3 +82,70 @@ export class BigQueryConfiguration implements IBigQueryConfiguration {
...
+    public set privateKey(_: string) {
+        config.noop();
+    }
+    public get privateKey(): string {
+        return config.noop();
+    }
+}
+
+@config.section
+export class PubSubPublisherConfiguration extends GcpAuthConfiguration
```
Comment (chandanRIT) [inline packages/gcp/src/config.ts:114]: shouldnt this implement IPubSubPublisherConfiguration

File: packages/gcp/src/PubSubSink.ts [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    constructor(private readonly config: IGcpAuthConfiguration & IPubSubPublisherConfiguration) {
+        this.logger = DefaultComponentContext.logger;
+        this.tracer = DefaultComponentContext.tracer;
+    }
+
+    public async initialize(ctx: IComponentContext): Promise<void> {
+        this.logger = ctx.logger;
+        this.tracer = ctx.tracer;
+        this.metrics = ctx.metrics;
+        this.producer = new PubSub({
```
Comment (chandanRIT) [inline packages/gcp/src/PubSubSink.ts:75]: producer can be initialized in the constructor , right ?

File: packages/gcp/src/PubSubSink.ts [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+    private emitMetrics(topic: string, eventType: string, result: PubSubMetricResults) {
+        this.metrics.increment(PubSubMetrics.MsgPublished, {
+            topic,
+            event_type: eventType,
+            result,
+        });
+    }
+
+    public async sink(output: IterableIterator<IPublishedMessage>): Promise<void> {
```
Comment (chandanRIT) [inline packages/gcp/src/PubSubSink.ts:100]: public before private

File: packages/gcp/src/PubSubSink.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+                    maxMilliseconds: this.config.maximumBatchWaitTime,
+                },
+            });
+            for (const message of messages) {
+                const span = this.tracer.startSpan(this.spanOperationName, {
+                    childOf: message.spanContext,
+                });
+                this.spanLogAndSetTags(span, this.sink.name, topic);
+                try {
+                    const messageId = await batchPublisher.publish(
```
Comment (chandanRIT) [inline packages/gcp/src/PubSubSink.ts:125]: looks like the messages are being published individually, 
isnt the intention to publish in batches ?

File: packages/gcp/src/PubSubSink.ts [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+                    span.log({ messageId });
+                    this.emitMetrics(
+                        topic,
+                        message.attributes[AttributeNames.eventType],
+                        PubSubMetricResults.Success
+                    );
+                    this.logger.debug("Message published to PubSub", { topic, messageId });
+                } catch (e) {
+                    failSpan(span, e);
+                    messages.forEach((message) =>
```
Comment (chandanRIT) [inline packages/gcp/src/PubSubSink.ts:138]: do we want to fail all the messages, if one of the messages fails to publish?

File: packages/gcp/src/config.ts (hunk 8 above)
Comment (k-gupta) [reply packages/gcp/src/config.ts:144]: Nope this is a byte conversion so you can write something like `1Mb`

File: packages/gcp/src/PubSubSink.ts (hunk 14 above)
Comment (k-gupta) [reply packages/gcp/src/PubSubSink.ts:122]: Ya its a batch publish

File: packages/gcp/src/__test__/PubSubSink.test.ts [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+
+    it("Reject on error from PubSub topic", async () => {
+        const testApp = createTestApp(messagesWithoutTopic, sink, ErrorHandlingMode.LogAndFail);
+        mockTopic.mockImplementation(() => {
+            throw err;
+        });
+        await expect(testApp).rejects.toThrowError();
+    });
+
+    it("Reject on error from PubSub topic", async () => {
```
Comment (sklose) [inline packages/gcp/src/__test__/PubSubSink.test.ts:156]: nit: can you change wording for the tests slightly so this reads like a sentence?

```
describe("PubSubSink", {
  it("writes to topic specified in metadata", async () => {

  });
}
```

Comment (k-gupta) [reply packages/gcp/src/__test__/PubSubSink.test.ts]: I thought I changed it. It should be resolved now

**Pull request merged**